riscv64: [run-make] tests/run-make/instrument-mcount-link-pg failed: undefined reference to `mcount'

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

[RVCI](https://community-ci.openruyi.cn/rvci/job/RustCI/job/RustCheck/992/) detected the issue some days ago:

```rust
--- stderr -------------------------------
command failed at line 12
LD_LIBRARY_PATH="/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/test/run-make/instrument-mcount-link-pg/rmake_out:/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib:/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/bootstrap-tools/riscv64gc-unknown-linux-gnu/release/deps:/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage0/lib/rustlib/riscv64gc-unknown-linux-gnu/lib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/bin/rustc" "-L" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/test/run-make/instrument-mcount-link-pg/rmake_out" "main.rs" "-Zinstrument-mcount" "--print" "link-args" "--target=riscv64gc-unknown-linux-gnu"
output status: `exit status: 1`
=== STDOUT ===
LC_ALL="C" PATH="/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/bin:/usr/lib/ccache:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" VSLANG="1033" "cc" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/test/run-make/instrument-mcount-link-pg/rmake_out/rustcmPlFPU/symbols.o" "main.main.d17f5fbe6225cf88-cgu.0.rcgu.o" "main.2uoctswmurc6ir5rvoay0p9ke.rcgu.o" "-Wl,--as-needed" "-Wl,-Bstatic" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libstd-48141712450b9943.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libpanic_unwind-3db3599c6071e7b4.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libobject-fb1936aa9a270a49.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libmemchr-d3641458914e6b02.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libaddr2line-550fecb5e18cbf8e.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libgimli-7b90e491d515fb03.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libcfg_if-1bf8868cc323b944.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/librustc_demangle-5af7e9ec8df6b802.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libstd_detect-407f69dccc3bb64d.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libhashbrown-784a66d2b4b6d55f.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/librustc_std_workspace_alloc-623bb70ec8dc8a24.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libminiz_oxide-dfae77dfd70734e1.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libadler2-d7690005a01fc3cb.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libunwind-c077c9f61c9b1b8f.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/liblibc-39e5c65298fdb844.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/librustc_std_workspace_core-b21702c5db53146f.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/liballoc-5ba63f3312586252.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libcore-87e2f8a2e5910157.rlib" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libcompiler_builtins-f0451092e26251ae.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/test/run-make/instrument-mcount-link-pg/rmake_out/rustcmPlFPU/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/test/run-make/instrument-mcount-link-pg/rmake_out" "-L" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/stage2/lib/rustlib/riscv64gc-unknown-linux-gnu/lib" "-o" "main" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs" "-pg"


=== STDERR ===
error: linking with `cc` failed: exit status: 1
  |
  = note:  "cc" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/test/run-make/instrument-mcount-link-pg/rmake_out/rustcmPlFPU/symbols.o" "<2 object files omitted>" "-Wl,--as-needed" "-Wl,-Bstatic" "<sysroot>/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/{libstd-*,libpanic_unwind-*,libobject-*,libmemchr-*,libaddr2line-*,libgimli-*,libcfg_if-*,librustc_demangle-*,libstd_detect-*,libhashbrown-*,librustc_std_workspace_alloc-*,libminiz_oxide-*,libadler2-*,libunwind-*,liblibc-*,librustc_std_workspace_core-*,liballoc-*,libcore-*,libcompiler_builtins-*}.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/test/run-make/instrument-mcount-link-pg/rmake_out/rustcmPlFPU/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/home/jenkins/rust/build/riscv64gc-unknown-linux-gnu/test/run-make/instrument-mcount-link-pg/rmake_out" "-L" "<sysroot>/lib/rustlib/riscv64gc-unknown-linux-gnu/lib" "-o" "main" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs" "-pg"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: /usr/bin/ld: main.main.d17f5fbe6225cf88-cgu.0.rcgu.o: in function `std::rt::lang_start::<()>':
          main.d17f5fbe6225cf88-cgu.0:(.text._RINvNtCsg5Ro7eDQFyR_3std2rt10lang_startuECshZ9kMv8z9bM_4main+0x22): undefined reference to `mcount'
          /usr/bin/ld: main.main.d17f5fbe6225cf88-cgu.0.rcgu.o: in function `std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>':
          main.d17f5fbe6225cf88-cgu.0:(.text._RINvNtNtCsg5Ro7eDQFyR_3std3sys9backtrace28___rust_begin_short_backtraceFEuuECshZ9kMv8z9bM_4main+0x16): undefined reference to `mcount'
          /usr/bin/ld: main.main.d17f5fbe6225cf88-cgu.0.rcgu.o: in function `std::rt::lang_start::<()>::{closure#0}':
          main.d17f5fbe6225cf88-cgu.0:(.text._RNCINvNtCsg5Ro7eDQFyR_3std2rt10lang_startuE0CshZ9kMv8z9bM_4main+0x16): undefined reference to `mcount'
          /usr/bin/ld: main.main.d17f5fbe6225cf88-cgu.0.rcgu.o: in function `<std::rt::lang_start<()>::{closure#0} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}':
          main.d17f5fbe6225cf88-cgu.0:(.text._RNSNvYNCINvNtCsg5Ro7eDQFyR_3std2rt10lang_startuE0INtNtNtCskYCn7n3ooyn_4core3ops8function6FnOnceuE9call_once6vtableCshZ9kMv8z9bM_4main+0x16): undefined reference to `mcount'
          /usr/bin/ld: main.main.d17f5fbe6225cf88-cgu.0.rcgu.o: in function `main::main':
          main.d17f5fbe6225cf88-cgu.0:(.text._RNvCshZ9kMv8z9bM_4main4main+0xa): undefined reference to `mcount'
          /usr/bin/ld: main.main.d17f5fbe6225cf88-cgu.0.rcgu.o:main.d17f5fbe6225cf88-cgu.0:(.text._RNvMs4_NtCskYCn7n3ooyn_4core3fmtNtB5_9Arguments8from_strCshZ9kMv8z9bM_4main+0x1a): more undefined references to `mcount' follow
          collect2: error: ld returned 1 exit status
```

Opening the issue to track it.

--- a/compiler/rustc_target/src/spec/targets/riscv32gc_unknown_linux_gnu.rs
+++ b/compiler/rustc_target/src/spec/targets/riscv32gc_unknown_linux_gnu.rs
@@ -23,6 +23,7 @@ pub(crate) fn target() -> Target {
             llvm_abiname: LlvmAbi::Ilp32d,
             max_atomic_width: Some(32),
             supported_split_debuginfo: Cow::Borrowed(&[SplitDebuginfo::Off]),
+            mcount: "\u{1}_mcount".into(),
             ..base::linux_gnu::opts()
         },
     }

--- a/compiler/rustc_target/src/spec/targets/riscv64a23_unknown_linux_gnu.rs
+++ b/compiler/rustc_target/src/spec/targets/riscv64a23_unknown_linux_gnu.rs
@@ -23,6 +23,7 @@ pub(crate) fn target() -> Target {
             llvm_abiname: LlvmAbi::Lp64d,
             max_atomic_width: Some(64),
             supported_split_debuginfo: Cow::Borrowed(&[SplitDebuginfo::Off]),
+            mcount: "\u{1}_mcount".into(),
             ..base::linux_gnu::opts()
         },
     }

--- a/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_linux_gnu.rs
+++ b/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_linux_gnu.rs
@@ -23,6 +23,7 @@ pub(crate) fn target() -> Target {
             llvm_abiname: LlvmAbi::Lp64d,
             max_atomic_width: Some(64),
             supported_split_debuginfo: Cow::Borrowed(&[SplitDebuginfo::Off]),
+            mcount: "\u{1}_mcount".into(),
             ..base::linux_gnu::opts()
         },
     }

